server/container/util-java8/src/main/java/org/apache/james/util/OptionalUtils.java

```diff
@@ -45,4 +45,17 @@
             .filter(Optional::isPresent)
             .orElse(optional2);
     }
+
+    public static <T> boolean containsDifferent(Optional<T> requestValue, T storeValue) {
+        return requestValue
+            .filter(value -> !value.equals(storeValue))
+            .isPresent();
+    }
+
+    public static <T> boolean contentChangedCompareToStoreValue(Optional<T> requestValue, Optional<T> storeValue) {
```

Comment:
I don't see how is this method helpful :  `!requestValue.equals(storeValue)` does the same, isn't it ?

Reply:
Not really because it will always return false if first argument is empty whereas empty.equals()

Reply:
The function name is "contentChanged ..." and "!" is mean the content is changed. WDYT?

Reply:
The trick is that requestValue = Optional.empty() is not supposed to change stored value, even when present, so equals is sadly not enought... (Yes I first thought like you)

Reply:
I didn't understand your comment, i'm sorry

Reply:
|---------------------|--------------------|--------------------|
|RequestedValue|StoredValue|Is the data changed|
|empty|empty|No|
|empty|of(X)|No ! This is what makes this not equivalent to `different`|
|of(X)|empty|Yes|
|of(X)|of(Y)|Yes|
|of(X)|of(X)|No|

Reply:
Ok, it's a very specific rule actually. Please don't trick Optional for domain rules. You can very well have this method, but not in OptionalUtils

Reply:
Changed